### ISSUE TITLE
Unmerged reads with short identical overlap

### ISSUE THREAD

**Issue by andrefaure** [torognes/vsearch#436]
Hi, I am obtaining unexpected results from vsearch -fastq_mergepairs when there is a short overlapping region with no mismatches. For example merging fails on the following two reads which are simply the reverse complement of each other:

read1.fastq:
```
@SRR13267797.1
AACTTCGC
+
FFFFFFFF
```

read2.fastq:
```
@SRR13267797.1
GCGAAGTT
+
FFFFFFFF
```

when running vsearch as follows:
`vsearch -fastq_mergepairs read1.fastq --reverse read2.fastq --fastqout output.fastq --fastq_minlen 4 --fastq_minovlen 4 --fastq_maxdiffs 4`

The reason given is "alignment score too low, or score drop too high". Why do these reads fail to merge?

I am using VSEARCH v2.14.2_linux_x86_64.

Thanks for your help!

**Comment by frederic-mahe** [comment]
Hi @andrefaure I think you are right. I expect the following toy-example to produce a merged sequence: 
```sh
vsearch \
    --fastq_mergepairs <(printf "@s1\nAACTTCGC\n+\nFFFFFFFF\n") \
    --reverse <(printf "@s2\nGCGAAGTT\n+\nFFFFFFFF\n") \
    --fastq_minovlen 4 \
    --fastqout -
```
but merging is rejected:
```
vsearch v2.15.1_linux_x86_64, 62.8GB RAM, 8 cores
https://github.com/torognes/vsearch

Merging reads 100%
         1  Pairs
         0  Merged (0.0%)
         1  Not merged (100.0%)

Pairs that failed merging due to various reasons:
         1  alignment score too low, or score drop too high

Statistics of all reads:
      8.00  Mean read length
```

These options should not change the results `--fastq_minlen` (default is 1), `--fastq_minmergelen` (default is 1), and `--fastq_maxdiffs` (default is 10).

**Comment by andrefaure** [comment]
Thanks @frederic-mahe, so is this a bug? Or is there some way to ensure vsearch merges such pairs as expected?

**Comment by torognes** [comment]
Hi @andrefaure, thanks for reporting this issue.

There are some additional heuristics that vsearch applies when it tries to merge reads that are not described in detail in the documentation and some parameters that cannot be changed. As part of this, it computes an alignment score of the two overlapping sequences, and requires this alignment score to be above a certain threshold. That alignment score depends on whether the sequences match exactly or not and on the quality score of the bases. To obtain this minimum score it actually requires at least 9 matching bases. There is currently no option to adjust this. At the moment it does not help to set `--fastq_minovlen` to anything less than 9. A warning or error should probably have been given when trying to set some parameters below this limit.

There some other heuristics that try to avoid alignments with stretches of mismatching bases, potential repeats etc. An issue (#430) about this has been raised recently.

When I get time, I'll try to improve the algorithm and document it better.

**Comment by andrefaure** [comment]
Ok I understand - thanks a lot for the reply and explanation @torognes ! In the meantime one shouldn't attempt alignment of reads with less than 9 matching bases then.

**Comment by torognes** [comment]
That's right. Thank you.

**Comment by frederic-mahe** [comment]
Minimal working example (9 nt):
```sh
vsearch \
    --fastq_mergepairs <(printf "@s1\nGAACTTCGC\n+\nFFFFFFFFF\n") \
    --reverse <(printf "@s2\nGCGAAGTTC\n+\nFFFFFFFFF\n") \
    --fastq_minovlen 4 \
    --fastqout -
```

**Comment by torognes** [comment]
In commit 6f39ea95e499b93fbb80eb24c4f59dbada7cc4df, I have made a change to the `fastq_mergepairs` command. It now complains if the argument to the `--fastq_minovlen` option is less than 5. If the argument is less than 10 it adjust some parameters that makes vsearch successfully merge sequences with shorter overlaps than 9. It is now possible to merge sequences with overlaps as short as 5 bp if they are perfect. It should solve the issue. I've also made other changes to the command (see issue #430).
